spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1300,6 +1300,82 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
+
+==== Discard fragment
+A fragment in the path is indicated by the first occurrence of a `\#` character.  Any `#` character and following fragment is removed from the path and discarded.
+
+==== Decoding of non-special characters
+Characters other than `/`, `;` and `%` that are encoded in `%nn` form are decoded and the resulting octet sequences is treated as UTF-8 and converted to a character sequence.
+
+> Note that special characters cannot be part of a UTF-8 character sequence as all such sequences are comprised of negative octets.
+
+> Note this is not reserved characters as defined by RFC3986, as that does not include `%` and includes many characters we don't care about. Avoiding a second decoding is worthwhile.
+
+==== Collapse sequences of multiple `"/"` characters
+> **WARNING** Swapping the order of stage 3 and stage 4 may be significant. Consider `"/aaa/bbb//../"`.
+
+> **TODO** Are we sure we don't want to do this in the other order?
+
+Any sequence of more than one `"/"` character in the URI must be replaced with a single `"/"`.
+
+==== Remove dot-segments+
+* A path not starting with "/" must be rejected with a 400 response.
+* Sequences of the form `"/./"` must be replaced with `"/"`.
+* Sequences of the form `"/" + segment + "/../"` must be replaced with `"/"`.
+* If there is no preceding segment for a `".."` segment then return a 400 response.
+
+==== Removal of path parameters
+A path segment containing the `";"` character is split at the first occurence of `";"`. The segment is replaced by the character sequence preceeding the `";"`. The characters following the `";"` are considered a path parameters and may be preserved by the container for later processing (eg `jsessionid`).
+
+> TODO How do we handle URIs like `/foo/;/bar`?  I think as currently written we end up with `/foo//bar` ?
+
+==== Decoding of remaining `%nn` sequences
+Any remaining `%nn` sequences in the path should be decoded. Some containers may be configured to leave some specific characters encoded (eg. the characters '/' and '%' may be left decoded by some container configuration).
+
+==== Mapping URI to context and resource
+The decoded path is used to map the request to a context and resource within the context. This form of the URI path is used for all subsequent mapping (web applications, servlet, filters and security constraints).
+
+==== Rejecting Suspicious Sequences
+If suspicious sequences are discovered during the prior steps, the request must be rejected with a 400 bad request using the error handling of the matched context.  By default the set of suspicious sequences includes:
+
+ * The encoded `"/"` character
+ * Any `"."` or `".."` segment that had a path parameter
+ * Any `"."` or `".."` segment with any encoded characters
+ * The `"\"` character encoded or not.
+ * Any control characters either encoded or not.
```

Comment:
Do we want this? If so, how do we define control characters?  < 0x20? Is 0x7F (DEL) OK?

Reply:
I'm not sure we need this. Their presence would be unusual but I don't see a good reason to exclude them by default at this point. (I'll note that if the URIs gets passed to the default servlet it might reject them with a 404 if there character is not a legal file name character for the underlying file system).

Reply:
I'm kind of concerned about characters like `%00` that can fool any native code handling.
Also  `%08`, `%0D` and `%7F` can be used to hide characters in a displayed URI (backspace, carriage ret and delete).

Perhaps if we don't reject these by default, we use this as an example of extra rejections that a container might do?

Reply:
I am ok with rejecting by default. I think the actual use cases for this are pretty much non-existent, so if they are present there is almost certainly something nefarious going on.